### ISSUE TITLE
Differing behavior between printers, when removing one value from an annotation type element

### ISSUE THREAD

**Issue by jbmestelan** [javaparser/javaparser#2936]
Hello, and thanks for JavaParser.

When modifying an annotation, and removing one value among a list in the annotation type element, I find that this has no effect on its lexically printed representation. The compilation unit has taken the change into account, but it seems that it will just print unchanged. 

Conversely, the pretty print will reflect those changes.

You can find the code to reproduce [here](https://github.com/Mestelan/JavaParser_report/tree/main/src/test/java). Is there a better way to achieve this?

**Comment by MysterAitch** [comment]
> I find that this has no effect on its lexically printed representation. 

What does this mean? Could you give an example of the actual output and your desired output please?

**Comment by jbmestelan** [comment]
Yes, sorry. 

Starting from this class:
```java
@MyAnnotation(myElements = {"good", "bad", "ugly"})
public final class MyClass {
}
```

and after the manipulations described [here](https://github.com/Mestelan/JavaParser_report/blob/main/src/main/java/AnnotationRemoverVisitor.java), which remove the `"bad"` value, we get this output:

```
Using pretty print:
@MyAnnotation(myElements = { "good", "ugly" })
public final class MyClass {
}


Using lexical preserving printer:
@MyAnnotation(myElements = {"good", "bad", "ugly"})
public final class MyClass {
}
```

I expect the value to be removed, whichever way is used to print back the result of the manipulation. Am I doing the transformation in a unexpected way?

```java
final class AnnotationRemoverVisitor extends ModifierVisitor<Void> {
    public static final String REFERENCE_TO_BE_DELETED = "bad";

    public AnnotationExpr visit(final NormalAnnotationExpr annotation, final Void arg) {
        super.visit(annotation, arg);

        MemberValuePair mvp = annotation.getPairs().get(0);
        Expression value = mvp.getValue();
        if ((value instanceof ArrayInitializerExpr)) {
            NodeList<Expression> myElements = ((ArrayInitializerExpr) value).getValues();

            for (Iterator<Expression> iterator = myElements.iterator(); iterator.hasNext(); ) {
                Node elt = iterator.next();
                {
                    String nameAsString = ((StringLiteralExpr) elt).asString();
                    if (REFERENCE_TO_BE_DELETED.equals(nameAsString))
                        iterator.remove();
                }
            }
            if (myElements.size() == 0)
                return null;
            return annotation;
        }
    }
}
```

**Comment by MysterAitch** [comment]
Cheers - that makes more sense, thanks! (was a bit tricky navigating multiple files in nested directories using just the github UI!)

The only thing that jumps out is using `iterator.remove()` -- I'm not entirely sure whether that is effective, or whether the `Node#remove()` must be used (I don't use the visitor side of JavaParser). 

Edit: Dipping into the source code for `NodeList.java`, the list's remove method triggers the notifications (e.g. to the LPP observer) that an element was removed. 

```java
    // NodeList.java
    public boolean remove(Node node) {
        int index = innerList.indexOf(node);
        if (index != -1) {
            notifyElementRemoved(index, node);
            node.setParentNode(null);
        }
        return innerList.remove(node);
    }
```

Looking further down the file, I also see this `NodeList#iterator()` method which shows that we are just delegating it to `List.iterator()`, thus are bypassing the remove method above (hence the "TODO"). 

```java
    // NodeList.java
    @Override
    public Iterator<N> iterator() {
        // TODO take care of "Iterator.remove"
        return innerList.iterator();
    }
```

**Comment by jbmestelan** [comment]
Thank you. It is right that, if I replace the call to `iterator.remove()` with `Node#remove()`, the test has the expected behaviour!

Here is the transformation, rewritten to directly manipulate the nodes:

This way of deleting, using `Iterator.remove`, causes the problem:
```java
    private void removeNode(CompilationUnit compilationUnit) {
        TypeDeclaration<?> type = compilationUnit.getType(0);
        NormalAnnotationExpr annotation = (NormalAnnotationExpr) type.getAnnotationByName("MyAnnotation").orElseThrow(IllegalStateException::new);
        MemberValuePair mvp = annotation.getPairs().get(0);
        Expression value = mvp.getValue();

        if (!(value instanceof ArrayInitializerExpr))
            throw new IllegalStateException();

        NodeList<Expression> myElements = ((ArrayInitializerExpr) value).getValues();
        for (Iterator<Expression> iterator = myElements.iterator(); iterator.hasNext(); ) {
            Node elt = iterator.next();
            {
                String nameAsString = ((StringLiteralExpr) elt).asString();
                if (AnnotationRemoverVisitor.REFERENCE_TO_BE_DELETED.equals(nameAsString)) {
                    iterator.remove();
                }
            }
        }
    }
```
Result
```
Using pretty print:
@MyAnnotation(myElements = { "good", "ugly" })
public final class MyClass {
}


Using lexical preserving printer:
@MyAnnotation(myElements = {"good", "bad", "ugly"})
public final class MyClass {
}
```

This other way, using `Node#remove()`, does not cause the problem :
```java
    private void removeNode(CompilationUnit compilationUnit) {
        TypeDeclaration<?> type = compilationUnit.getType(0);
        NormalAnnotationExpr annotation = (NormalAnnotationExpr) type.getAnnotationByName("MyAnnotation").orElseThrow(IllegalStateException::new);
        MemberValuePair mvp = annotation.getPairs().get(0);
        Expression value = mvp.getValue();

        if (!(value instanceof ArrayInitializerExpr))
            throw new IllegalStateException();

        NodeList<Expression> myElements = ((ArrayInitializerExpr) value).getValues();
        for (Iterator<Expression> iterator = myElements.iterator(); iterator.hasNext(); ) {
            Node elt = iterator.next();
            {
                String nameAsString = ((StringLiteralExpr) elt).asString();
                if (AnnotationRemoverVisitor.REFERENCE_TO_BE_DELETED.equals(nameAsString)) {
                    myElements.remove(elt);
                    break;
                }
            }
        }
    }
```
Result
```
Using pretty print:
@MyAnnotation(myElements = { "good", "ugly" })
public final class MyClass {
}


Using lexical preserving printer:
@MyAnnotation(myElements = {"good", "ugly"})
public final class MyClass {
}

```
And indeed, I just spotted the `TODO` now. What was puzzling me was that the compilation unit seemed impacted by the changes, but they did not get serialized; well, they did not get serialized when using lexical preserving print.

**Comment by jbmestelan** [comment]
Now, knowing this, we can work around this behaviour, and find alternate ways to iterate over the nodes to selectively remove some items. I have used a `break` here, which works in my use-case, because each value should occur only once. Or I suppose we could delete the elements over several loops...

**Comment by jbmestelan** [comment]
But maybe the `NodeList#iterator` should return an implementation which fails fast when its `remove` method is invoked?

**Comment by MysterAitch** [comment]
Personally, my default approach would be to use the `findAll()` method and stream filters to collect a list of nodes to remove, then iterate over that list calling `Node#remove()`.

Again, not sure if that'll work with the visitors but I don't immediately see why it wouldn't.

I'm currently looking at fixing this with a custom iterator. If I don't get anywhere fast with that (maybe it's not as simple as I hope it will be) then, yes, the todo/warning about using the iterator with LPP enabled will be made more prominent :+1:


Thanks for the report!! :)

**Comment by MysterAitch** [comment]
I've just pushed a custom iterator implementation that delegates to the arraylist's iterator implementation. It'll be merged once the CI is complete (travis delays.. #2919).

As far as I can tell the spliterator doesn't need to be done too because it doesn't allow modification, but I might be mistaken (please do correct me!). If it does, that can be added as a separate issue/PR though.
